pkg/snmp/traps/formatter.go

```diff
@@ -0,0 +1,249 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package traps
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/gosnmp/gosnmp"
+)
+
+// Formatter is an interface to extract and format raw SNMP Traps
+type Formatter interface {
+	FormatPacket(packet *SnmpPacket) ([]byte, error)
+	GetTags(packet *SnmpPacket) []string
+}
+
+// JSONFormatter is a Formatter implementation that transforms Traps into JSON
+type JSONFormatter struct {
+	oidResolver OIDResolver
+	namespace   string
+}
+
+type trapVariable struct {
+	OID     string      `json:"oid"`
+	VarType string      `json:"type"`
+	Value   interface{} `json:"value"`
+}
+
+const (
+	sysUpTimeInstanceOID = "1.3.6.1.2.1.1.3.0"
+	snmpTrapOID          = "1.3.6.1.6.3.1.1.4.1.0"
+)
+
+// NewJSONFormatter creates a new JSONFormatter instance with an optional OIDResolver variable.
+func NewJSONFormatter(oidResolver OIDResolver) (JSONFormatter, error) {
+	if oidResolver == nil {
+		return JSONFormatter{}, fmt.Errorf("NewJSONFormatter called with a nil OIDResolver")
+	}
+	namespace := GetNamespace()
+	return JSONFormatter{oidResolver, namespace}, nil
+}
+
+// FormatPacket converts a raw SNMP trap packet to a FormattedSnmpPacket containing the JSON data and the tags to attach
+func (f JSONFormatter) FormatPacket(packet *SnmpPacket) ([]byte, error) {
+	var formattedData map[string]interface{}
+	var err error
+	if packet.Content.Version == gosnmp.Version1 {
+		formattedData = f.formatV1Trap(packet.Content)
+	} else {
+		formattedData, err = f.formatTrap(packet.Content)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return json.Marshal(formattedData)
+}
+
+// GetTags returns a list of tags associated to an SNMP trap packet.
+func (f JSONFormatter) GetTags(packet *SnmpPacket) []string {
+	return []string{
+		"snmp_version:" + formatVersion(packet.Content),
+		"device_namespace:" + f.namespace,
+		"snmp_device:" + packet.Addr.IP.String(),
+	}
+}
+
+func (f JSONFormatter) formatV1Trap(packet *gosnmp.SnmpPacket) map[string]interface{} {
+	data := make(map[string]interface{})
+	data["uptime"] = uint32(packet.Timestamp)
+	enterpriseOid := NormalizeOID(packet.Enterprise)
+	genericTrap := packet.GenericTrap
+	specificTrap := packet.SpecificTrap
+	var trapOID string
+	if genericTrap == 6 {
+		// Vendor-specific trap
+		trapOID = fmt.Sprintf("%s.0.%d", enterpriseOid, specificTrap)
+	} else {
+		// Generic trap
+		trapOID = fmt.Sprintf("%s.%d", genericTrapOid, genericTrap+1)
+	}
+	data["snmpTrapOID"] = trapOID
+	trapMetadata, err := f.oidResolver.GetTrapMetadata(trapOID)
+	if err != nil {
+		log.Debugf("unable to resolve OID: %s", err)
+	} else {
+		data["snmpTrapName"] = trapMetadata.Name
```

Comment:
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**🥜 nitpick:** `snmpTrap` is here in CamelCase while `variables_raw` below uses train_case. Is it intentional ?‏

Reply:
Changing `variables_raw` to `variables` actually, we don't need `raw` anymore as there is no conflict anymore